### PR TITLE
Design for mon failover

### PR DESCRIPTION
**Pull request by travisn**
A design doc for healthy mon quorum and failover. [skip ci]

### CODE REVIEW

File: design/mon-health.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+The Rook operator creates a mon with a ReplicaSet to ensure that the mon pod will always be restarted if it fails. If a mon pod stops for any reason, Kubernetes will 
+automatically start the pod up again.
+
+In order for a mon to support a pod/node restart, you will need to set the `dataDirHostPath` so the mon metadata will be persisted to disk in this folder. 
+This will allow the mon to start back up with its existing metadata and continue where it left off even if the pod had to be re-created. Without this host path, the mon cannot
+start again.
+
+## Failing over a Monitor
+If the mitigating steps for mon failure don't bring a mon back up after some period of time (five minutes), the operator will make the decision to terminate the old monitor pod
```
Comment (galexrt) [inline design/mon-health.md]: I think it is worth mentioning that the mon back up time is configurable by a flag.

File: design/mon-health.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+## Monitor Quorum
+Multiple mons work together to provide redundancy by each keeping a copy of the metadata. A variation of the distributed algorithm Paxos is used to establish 
+consensus about the state of the cluster. Paxos requires a super-majority of mons to be running in order to establish quorum and perform operations in the cluster. 
+If the majority of mons are not running, quorum is lost and nothing can be done in the cluster. 
+
+### How many mons?
+Most commonly a cluster will have three mons. This would mean that one mon could go down and allow the cluster to remain healthy. 
+You would still have 2/3 mons running to give you consensus in the cluster for any operation.
+
+You will always want an odd number of mons. Fifty percent of mons will not be sufficient to maintain quorum. Ff you had two mons and one of them went down,
```
Comment (galexrt) [inline design/mon-health.md]: As you wrote in the private chat, the "Ff" should be "If".

**Comment by travisn** [comment]
@galexrt updated per your suggestions, PTAL

**Comment by galexrt** [comment]
LGTM

File: Documentation/cluster-crd.md [hunk 3]
```diff
@@ -26,7 +26,8 @@ Settings can be specified at the global level to apply to the cluster as a whole
   - **WARNING**: For test scenarios, if you delete a cluster and start a new cluster on the same hosts, the path used by `dataDirHostPath` must be deleted. Otherwise, stale keys and other config will remain from the previous cluster and the new mons will fail to start. 
 If this value is empty, each pod will get an ephemeral directory to store their config files that is tied to the lifetime of the pod running on that node. More details can be found in the Kubernetes [empty dir docs](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir).
 - `hostNetwork`: uses network of the hosts instead of using the SDN below the containers.
-- `monCount`: set the amount of mons to be started. The number must be odd and between `1` and `9`. Default if not specified is `3`.
+- `monCount`: set the number of mons to be started. The number should be odd and between `1` and `9`. Default if not specified is `3`.
+For more details on the mons and when to choose a number other than `3`, see the [mon health design doc](https://github.com/rook/rook/blob/master/design/mon-health.md).
```
Comment (jbw976) [inline Documentation/cluster-crd.md]: can we use a relative link here instead?

Comment (travisn) [reply Documentation/cluster-crd.md]: i don't know how the rook.io docs can reference something on github.com with a relative link.

**Comment by galexrt** [comment]
There may be one thing that is missing from the current logic.
I don't think that currently a decreased `monCount` will cause the health check to remove mons.
**Example**:
Current behavior:
Cluster with `monCount: 5`, 5 mons exist.
Cluster CRD updated to `monCount: 3`, currently the mons would not be removed.

Idea behavior:
Cluster with `monCount: 5`, 5 mons exist.
Cluster CRD updated to `monCount: 3`, health check removes one by one mon until three mons are left.

The removal of mons should be done by the health check I would say.

**Comment by travisn** [comment]
@galexrt agreed, the health check sounds like the right place for that change. Want to open a separate PR for that? I'll go ahead and merge this, thanks

**Comment by galexrt** [comment]
An issue to add/change this behavior exists, see https://github.com/rook/rook/issues/1193.

**Pull request merged**